### ISSUE TITLE
Support state selection natively in the Dagster framework

### ISSUE THREAD

**Issue by rexledesma** [dagster-io/dagster#15872]
### What's the use case?

dbt has [the concept of state](https://docs.getdbt.com/reference/node-selection/methods#the-state-method), which allows for selection of dbt nodes based on whether the node is new or modified.

Some use cases:

- `state:new`: I can select all new models and run a dbt command on them
- `state:modified`: I can select all new models, and any models with changes and run a dbt command on them.
- `state:modified.XXX`: I can select all new models, and any models with a subset of the `state:modified` criteria (e.g. `modified.body`, `modified.config`, etc).
- When making a pull request, I can run only the modified data assets and their downstream dependencies in staging schema (e.g. `dbt build --select state:modified+1`)
- After merging a pull request, I can run only the modified data assets and their downstream dependencies in the production schema using a custom command (e.g. `dbt build --select state:modified+ --full-refresh`)

To use this in dbt, a reference to the previous manifest file must be given. In Dagster, we should be able to determine whether a given asset is modified or new, using its history from the asset catalog along with its current code version.

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by jackwillisupside** [comment]
+1 on this!!! This would be so helpful.

**Comment by jackwillisupside** [comment]
@rexledesma Is the only way you could do this currently by using some combo of context.instance.get_latest_materialization_code_versions(asset_keys) and another function that returns the current "next" code version and then selecting those asset keys and all downstream assets?

**Comment by sryza** [comment]
@jackwillisupside yes, that's the path I'd recommend exploring. Here's a snippet that might be a starting point:

```python
    definition_code_versions = context.defs.code_versions_by_asset_key
    latest_materialization_code_versions = (
        context.instance.get_latest_materialization_code_versions(definition_code_versions)
    )
    new_code_version_asset_keys = [
        key
        for key, def_code_version in definition_code_versions.items()
        if def_code_version != latest_materialization_code_versions.get(key)
    ]

    new_code_version_asset_keys = dict(
        definition_code_versions.items() - latest_materialization_code_versions.items()
    ).keys()
```
